### PR TITLE
[DAT-16006] Fixing v8 Checksum calculation issue when using sql dbms attribute

### LINKED ISSUE #4457: Using 4.23.0, my app fails to start because of invalid checksums
### Search first

- [X] I searched and no similar issues were found

### Description

Our application (Java/Spring/MySQL) is currently using 4.20.0 and works fine. When I upgrade to 4.23.0, it fails to start due to Liquibase failing due to invalid checksums.

```
	Caused by: liquibase.exception.CommandExecutionException: liquibase.exception.ValidationFailedException: Validation Failed:
     1 changesets check sum
          org/foo/db/phase5/foo-bar-2017-09-14.xml::DECREASE_FP_FIELD_FOR_PDP::jdoe was: 8:83871094f06fbeba7a3755bb6e7980c0 but is now: 8:7afe9a1d89ce253fbca2d513dcdc2834
```

### Steps To Reproduce

The offending changeset (names changed to protect the innocent):

```xml
    <changeSet author="jdoe" id="DECREASE_FP_FIELD_FOR_PDP" context="bootstrap,demo,unit">
        <comment>Words words words. Please review if your institution would like this change. See JIRA-123 for details.</comment>

        <sql dbms="mysql">
            update FP_DV_PAYEE_DTL_T A join FP_DV_PAYEE_DTL_T B ON A.FDOC_NBR = B.FDOC_NBR set A.DV_RMT_PRSN_NM = SUBSTR(B.DV_RMT_PRSN_NM,1,75);
        </sql>

        <sql dbms="oracle">
            update FP_DV_PAYEE_DTL_T A set DV_RMT_PRSN_NM = (select SUBSTR(DV_RMT_PRSN_NM,1,75) from FP_DV_PAYEE_DTL_T B where A.FDOC_NBR = B.FDOC_NBR);
        </sql>

        <modifyDataType columnName="DV_RMT_PRSN_NM" newDataType="VARCHAR(75)" tableName="FP_DV_PAYEE_DTL_T"/>
    </changeSet>
```

### Expected/Desired Behavior

My app would startup like it does with 4.20.0.

### Liquibase Version

4.23.0

### Database Vendor & Version

MySQL 5.7

### Liquibase Integration

Spring (not Boot)

### Liquibase Extensions

_No response_

### OS and/or Infrastructure Type/Provider

macOS

### Additional Context

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

### LINKED ISSUE #4665: dbms-specific changes cause checksum mismatch for unknown reasons on 4.21.0+
### Search first

- [X] I searched and no similar issues were found

### Description

Similar to #4156 and #4457, checksum calculations seem broken for me on multiple changeSets in `4.23.1`, `4.23.2` (and `4.21.1` although I skipped that due to #4156). While similar to these earlier issues, they seem different.

```
Caused by: liquibase.exception.CommandExecutionException: liquibase.exception.ValidationFailedException: Validation Failed:
     2 changesets check sum
          db-migration-scripts/initial/initialize.xml::0::gocd(generated) was: 8:84b408bb6848cef405f3e4420abfe5ac but is now: 8:d41d8cd98f00b204e9800998ecf8427e
          db-migration-scripts/initial/create-trigger.xml::107::gocd(generated) was: 8:126d525070c1cffacf1061fc4f021a65 but is now: 8:9790c3e9e279b99a5c4c96139fabe72b
```

I've no idea what the cause is, but two different files break.

https://github.com/gocd/gocd/blob/0f58107c851cf2df6ce7c6902eebde796dc1f742/db-support/db-migration/src/main/resources/db-migration-scripts/initial/initialize.xml#L18-L27
```xml
<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.8.xsd">
    <changeSet author="gocd(generated)" id="0">
        <sql dbms="postgresql">
            CREATE EXTENSION IF NOT EXISTS citext;
        </sql>
        <comment>This will load the case-insensitive text extension on PG.</comment>
    </changeSet>
</databaseChangeLog>
```

https://github.com/gocd/gocd/blob/0f58107c851cf2df6ce7c6902eebde796dc1f742/db-support/db-migration/src/main/resources/db-migration-scripts/initial/create-trigger.xml#L18-L58
```xml
<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.8.xsd">
    <changeSet author="gocd(generated)" id="107">
        <!-- prevent liquibase from splitting the statements on `;` -->
        <sql dbms="postgresql" splitStatements="false">
            <!-- @formatter:off -->
            CREATE FUNCTION update_stages_lastTransitionedTime() RETURNS TRIGGER AS $$
                BEGIN
                    UPDATE stages SET lastTransitionedTime = NEW.statechangetime WHERE stages.id = NEW.stageid;
                    RETURN NEW;
                END;
            $$ LANGUAGE plpgsql;

            CREATE TRIGGER lastTransitionedTimeUpdate
                AFTER INSERT ON buildStateTransitions
                FOR EACH ROW
                EXECUTE PROCEDURE update_stages_lastTransitionedTime();
            <!-- @formatter:on -->
        </sql>
        <sql dbms="h2">
            <!-- @formatter:off -->
            CREATE TRIGGER lastTransitionedTimeUpdate
                AFTER INSERT ON buildStateTransitions
                FOR EACH ROW
                CALL "com.thoughtworks.go.server.database.h2.UpdateLastTransitionedTime";
            <!-- @formatter:on -->
        </sql>
        <createProcedure dbms="mysql">
            <!-- @formatter:off -->
            CREATE TRIGGER lastTransitionedTimeUpdate
                AFTER INSERT ON buildStateTransitions
                FOR EACH ROW
                BEGIN
                    UPDATE stages SET lastTransitionedTime = NEW.statechangetime WHERE stages.id = NEW.stageid;
                END
            <!-- @formatter:on -->
        </createProcedure>
        <comment>Adds a trigger to update `stages.lastTransitionedTime`</comment>
    </changeSet>
</databaseChangeLog>
```

### Steps To Reproduce

- Run the migrations at https://github.com/gocd/gocd/tree/master/db-support/db-migration/src/main/resources/db-migration-scripts on Liquibase `4.20.0` (with an H2 database, although not sure that matters, as one error is for a postgresql only changeset).
- Then try again with `4.23.2`

I can possibly provide an easier reproducer (if this seems is not obviously a duplicate of another issue that I have missed).

### Expected/Desired Behavior

Should be able to upgrade existing database.

### Liquibase Version

4.21.1, 4.23.1, 4.23.2 (fail on all of these)

### Database Vendor & Version

H2 1.4.200

### Liquibase Integration

Direct java integration

### Liquibase Extensions

none

### OS and/or Infrastructure Type/Provider

MacOS Ventura

### Additional Context

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

### PR DESCRIPTION
**Pull request by filipelautert**
## Impact

- [X] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description

Checksums (v8) are calculated differently for XML changesets with the dbms attribute on the change itself (instead of the changeset) after 4.22.0 .

Fixes #4665
Fixes #4457

### CODE REVIEW

File: liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java [hunk 1]
```diff
@@ -731,20 +737,22 @@ public ExecType execute(DatabaseChangeLog databaseChangeLog, ChangeExecListener
 
                 log.fine("Reading ChangeSet: " + this);
                 for (Change change : getChanges()) {
-                    if (listener != null) {
-                        listener.willRun(change, this, changeLog, database);
-                    }
-                    if (change.generateStatementsVolatile(database)) {
-                        executor.comment("WARNING The following SQL may change each run and therefore is possibly incorrect and/or invalid:");
-                    }
+                    if ((!(change instanceof DbmsTargetedChange)) || DatabaseList.definitionMatches(((DbmsTargetedChange) change).getDbms(), database, true)) {
```
Comment (filipelautert) [inline liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java:740]: Moved this check back here (before 4.22.0 it was here and not at DbmsChangeSetFilter)

File: liquibase-standard/src/main/java/liquibase/changelog/filter/DbmsChangeSetFilter.java [hunk 2]
```diff
@@ -27,20 +27,12 @@ public ChangeSetFilterResult accepts(ChangeSet changeSet) {
             return new ChangeSetFilterResult(true, "No database connection, cannot evaluate dbms attribute", this.getClass(), getMdcName(), getDisplayName());
         }
         List<SqlVisitor> visitorsToRemove = new ArrayList<>();
-        List<Change> changesToRemove = new ArrayList<>();
         for (SqlVisitor visitor : changeSet.getSqlVisitors()) {
             if (!DatabaseList.definitionMatches(visitor.getApplicableDbms(), database, true)) {
                 visitorsToRemove.add(visitor);
             }
         }
-        for(Change change : changeSet.getChanges()){
```
Comment (filipelautert) [inline liquibase-standard/src/main/java/liquibase/changelog/filter/DbmsChangeSetFilter.java:36]: This filtering was moved back to Changeset class

File: liquibase-integration-tests/src/test/groovy/liquibase/command/core/UpgradeChecksumVersionIntegrationTest.groovy [hunk 3]
```diff
@@ -19,16 +19,16 @@ import spock.lang.Unroll
 @LiquibaseIntegrationTest
 class UpgradeChecksumVersionIntegrationTest extends Specification{
     @Shared
-    private DatabaseTestSystem mysql = (DatabaseTestSystem) Scope.getCurrentScope().getSingleton(TestSystemFactory.class).getTestSystem("mysql")
+    private DatabaseTestSystem h2 = (DatabaseTestSystem) Scope.getCurrentScope().getSingleton(TestSystemFactory.class).getTestSystem("h2")
```
Comment (filipelautert) [inline liquibase-integration-tests/src/test/groovy/liquibase/command/core/UpgradeChecksumVersionIntegrationTest.groovy:22]: We don't need mysql here as h2 is a lot faster to start.

**Review by wwillard7800 — APPROVED** [review]

**Comment by chadlwilson** [comment]
As noted in https://github.com/liquibase/liquibase/issues/4665#issuecomment-1741689649 this seems to resolve the issue in #4665 🙏

File: liquibase-integration-tests/src/test/groovy/liquibase/command/core/UpgradeChecksumVersionIntegrationTest.groovy (hunk 3 above)
Comment (StevenMassaro) [reply liquibase-integration-tests/src/test/groovy/liquibase/command/core/UpgradeChecksumVersionIntegrationTest.groovy:22]: I wonder how many other integration tests use a heavy database that could use H2. Probably most of them.

**Review by StevenMassaro — APPROVED** [review]

**Review by suryaaki2 — APPROVED** [review]

**Pull request merged**